File: src/no/omb/bigboat/io/HtmlReader.java
package no.omb.bigboat.io;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import no.omb.bigboat.BigBoat;
import no.omb.bigboat.data.Boat;
import no.omb.bigboat.data.RaceData;
import no.omb.bigboat.data.RaceData.DataFileType;
import no.omb.bigboat.data.RaceEntry;
import no.omb.bigboat.data.SeriesEntry;

public class HtmlReader implements DataReader {

	private static final HtmlReader instance = new HtmlReader();

	private static final String RACE_TABLE = "table.racetable";
	private static final String RACE_ROW = "tr.racerow";
	private static final String SUMMARY_TABLE = "table.summarytable";
	private static final String SUMMARY_ROW = "tr.summaryrow";

	private RaceData race;

	@Override
	public List<RaceEntry> parseDataFile(RaceData race) {
		this.race = race;
		return readDataFile();
	}

	private List<RaceEntry> readDataFile() {
		File input = new File(getFileName());
		try {
			Document doc = Jsoup.parse(input, "UTF-8", "http://example.com/");
			Elements tables = doc.select(isSummary() ? SUMMARY_TABLE : RACE_TABLE);
			return parseEntries(tables);
		} catch (IOException e) {
			e.printStackTrace();
		}
		return null;
	}
	
	private List<RaceEntry> parseEntries(Elements tables) {
		List<RaceEntry> raceEntries = new ArrayList<>();
		for (Element table : tables) {
			raceEntries.addAll(parseClassEntries(table));
		}
		Collections.sort(raceEntries);
		return raceEntries;
	}

	private List<RaceEntry> parseClassEntries(Element table) {
		List<RaceEntry> classEntries = new ArrayList<>();
		Elements rows = table.select(isSummary() ? SUMMARY_ROW : RACE_ROW);
		int classSize = rows.size();
		for (Element row : rows) {
			Elements cols = row.select("td");
			classEntries.add(parseEntry(cols, classSize));
		}		
		return classEntries;		
	}

	private RaceEntry parseEntry(Elements cols, int classSize) {
		RaceEntry entry = new RaceEntry();
		if (isSummary2()) {
			String place = cols.get(0).text();
			entry.setPlaceNo(Integer.parseInt(place.substring(0, place.length()-2)));
			entry.setBoat(new Boat(cols.get(2).text() + "-" + cols.get(3).text(),
					cols.get(4).text(),
					cols.get(5).text(),
					cols.get(6).text(),
					cols.get(7).text()));
		}
		else if (isSummary()) {
			String place = cols.get(0).text();
			entry.setPlaceNo(Integer.parseInt(place.substring(0, place.length()-2)));
			entry.setBoat(new Boat(cols.get(2).text().replace(' ', '-'),
					cols.get(3).text(),
					cols.get(4).text(),
					cols.get(5).text(),
					cols.get(6).text()));
		}
		else {
			entry.setPlaceNo(Integer.parseInt(cols.get(0).text()));
			entry.setBoat(new Boat(cols.get(2).text() + "-" + cols.get(3).text(),
					cols.get(5).text(),
					cols.get(4).text(),
					cols.get(6).text(),
					cols.get(7).text()));
		}
		entry.setClassSize(classSize);
		entry.setScore(entry.getPlaceNo() * SeriesEntry.SCORE_FACTOR * 1.0 / classSize);
		return entry;
	}

	public static HtmlReader getInstance() {
		return instance;
	}

	private String getFileName() {
		return BigBoat.DATA + "/" + race.getDataFileName() + ".html";
	}

	private boolean isSummary() {
		return race.getDataFileType() == DataFileType.SAILWAVE_SUMMARY_HTML || race.getDataFileType() == DataFileType.SAILWAVE_SUMMARY2_HTML;
	}

	private boolean isSummary2() {
		return race.getDataFileType() == DataFileType.SAILWAVE_SUMMARY2_HTML;
	}
	
}

File: src/no/omb/bigboat/io/HtmlReader.kt
package no.omb.bigboat.io

import no.omb.bigboat.BigBoat
import no.omb.bigboat.data.Boat
import no.omb.bigboat.data.RaceData
import no.omb.bigboat.data.RaceData.DataFileType
import no.omb.bigboat.data.RaceEntry
import no.omb.bigboat.data.SeriesEntry
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import org.jsoup.select.Elements
import java.io.*
import java.util.*

class HtmlReader : DataReader {
    private var race: RaceData? = null
    override fun parseDataFile(race: RaceData): List<RaceEntry> {
        this.race = race
        return readDataFile()
    }

    private fun readDataFile(): List<RaceEntry> {
        val input = File(fileName)
        try {
            val doc = Jsoup.parse(input, "UTF-8", "http://example.com/")
            val tables = doc.select(if (isSummary) SUMMARY_TABLE else RACE_TABLE)
            return parseEntries(tables)
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return emptyList()
    }

    private fun parseEntries(tables: Elements): List<RaceEntry> {
        val raceEntries: MutableList<RaceEntry> = ArrayList()
        for (table in tables) {
            raceEntries.addAll(parseClassEntries(table))
        }
        raceEntries.sort()
        return raceEntries
    }

    private fun parseClassEntries(table: Element): List<RaceEntry> {
        val classEntries: MutableList<RaceEntry> = ArrayList()
        val rows = table.select(if (isSummary) SUMMARY_ROW else RACE_ROW)
        val classSize = rows.size
        for (row in rows) {
            val cols = row.select("td")
            classEntries.add(parseEntry(cols, classSize))
        }
        return classEntries
    }

    private fun parseEntry(cols: Elements, classSize: Int): RaceEntry {
        val entry = RaceEntry()
        if (isSummary2) {
            val place = cols[0].text()
            entry.placeNo = place.substring(0, place.length - 2).toInt()
            entry.boat = Boat(
                cols[2].text() + "-" + cols[3].text(),
                cols[4].text(),
                cols[5].text(),
                cols[6].text(),
                cols[7].text()
            )
        } else if (isSummary) {
            val place = cols[0].text()
            entry.placeNo = place.substring(0, place.length - 2).toInt()
            entry.boat = Boat(
                cols[2].text().replace(' ', '-'),
                cols[3].text(),
                cols[4].text(),
                cols[5].text(),
                cols[6].text()
            )
        } else {
            entry.placeNo = cols[0].text().toInt()
            entry.boat = Boat(
                cols[2].text() + "-" + cols[3].text(),
                cols[5].text(),
                cols[4].text(),
                cols[6].text(),
                cols[7].text()
            )
        }
        entry.classSize = classSize
        entry.score = entry.placeNo * SeriesEntry.SCORE_FACTOR * 1.0 / classSize
        return entry
    }

    private val fileName: String
        get() = BigBoat.DATA + "/" + race!!.dataFileName + ".html"
    private val isSummary: Boolean
        get() = race!!.dataFileType == DataFileType.SAILWAVE_SUMMARY_HTML || race!!.dataFileType == DataFileType.SAILWAVE_SUMMARY2_HTML
    private val isSummary2: Boolean
        get() = race!!.dataFileType == DataFileType.SAILWAVE_SUMMARY2_HTML

    companion object {
        val instance = HtmlReader()
        private const val RACE_TABLE = "table.racetable"
        private const val RACE_ROW = "tr.racerow"
        private const val SUMMARY_TABLE = "table.summarytable"
        private const val SUMMARY_ROW = "tr.summaryrow"
    }
}
